### PR TITLE
[android] Remove weird static enum pattern from LocationHelper

### PR DESCRIPTION
**Pull request by rtsisyk**

### CODE REVIEW

File: android/app/src/main/java/app/organicmaps/routing/RoutingController.java [hunk 1]
```diff
@@ -266,15 +257,18 @@ public void attach(@NonNull Container container)
...
     mLastRouterType = Framework.nativeGetLastUsedRouter();
     mInvalidRoutePointsTransactionId = Framework.nativeInvalidRoutePointsTransactionId();
     mRemovingIntermediatePointsTransactionId = mInvalidRoutePointsTransactionId;
 
     Framework.nativeSetRoutingListener(mRoutingListener);
     Framework.nativeSetRouteProgressListener(mRoutingProgressListener);
-    Framework.nativeSetRoutingRecommendationListener(mRoutingRecommendationListener);
+    Framework.nativeSetRoutingRecommendationListener(recommendation -> UiThread.run(() -> {
+      if (recommendation == Framework.ROUTE_REBUILD_AFTER_POINTS_LOADING)
+        setStartPoint(LocationHelper.from(context).getMyPosition());
```
Comment (rtsisyk) [inline android/app/src/main/java/app/organicmaps/routing/RoutingController.java:270]: Does anyone know the purpose of RoutingRecommendationListener callback? This is the only controversal part in this patch...

Comment (biodranik) [reply android/app/src/main/java/app/organicmaps/routing/RoutingController.java:270]:
```
  if (m_routeRecommendCallback != nullptr)
  {
    // Check if we've found my position almost immediately after route points loading.
    auto constexpr kFoundLocationInterval = 2.0;
    auto const elapsed = chrono::steady_clock::now() - m_loadRoutePointsTimestamp;
    auto const sec = chrono::duration_cast<chrono::duration<double>>(elapsed).count();
    if (sec <= kFoundLocationInterval)
    {
      m_routeRecommendCallback(Recommendation::RebuildAfterPointsLoading);
      CancelRecommendation(Recommendation::RebuildAfterPointsLoading);
    }
  }
```

File: android/app/src/main/java/app/organicmaps/MwmApplication.java [hunk 2]
```diff
@@ -357,7 +367,7 @@ private void onBackground()
       Logger.i(LOCATION_TAG, "Navigation is in progress, keeping location in the background");
       return;
     }
-    LocationHelper.INSTANCE.stop();
+    LocationHelper.from(this).stop();
```
Comment (vng) [inline android/app/src/main/java/app/organicmaps/MwmApplication.java:370]: mLocationHelper?

File: android/app/src/main/java/app/organicmaps/routing/RoutingController.java (hunk 1 above)
Comment (vng) [reply android/app/src/main/java/app/organicmaps/routing/RoutingController.java:270]: What I see in the code:
MwmActivity.onSafeCreate->RestoreRouteTask.restoreRoute->LoadRoutePoints->m_loadRoutePointsTimestamp->SetUserCurrentPosition->m_routeRecommendCallback

I suppose the UX logic is: in navigation mode (some route exists), the app can close activity (go to background). When _foregrounding_ the app again, it loads a saved (previous) route and calls callback immediately after getting a location after loading the route.

File: android/app/src/main/java/app/organicmaps/MwmActivity.java [hunk 3]
```diff
@@ -1666,10 +1665,11 @@ public void onSuggestRebuildRoute()
       builder.setPositiveButton(R.string.ok, (dialog, which) -> RoutingController.get().swapPoints());
     else
     {
-      if (LocationHelper.INSTANCE.getMyPosition() == null)
+      final MapObject myPosition = LocationHelper.from(this).getMyPosition();
+      if (myPosition == null)
         builder.setMessage(null).setNegativeButton(null, null);
-
-      builder.setPositiveButton(R.string.ok, (dialog, which) -> RoutingController.get().setStartFromMyPosition());
+      else
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/MwmActivity.java:1671]: Moved to a separate PR https://github.com/organicmaps/organicmaps/pull/6052

File: android/app/src/main/java/app/organicmaps/routing/RoutingController.java (hunk 1 above)
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/routing/RoutingController.java:270]: ok, the current approach should work, but I want to remove static members and global RoutingController on the next step.

**Review by biodranik — APPROVED** [review]

**Review by vng — APPROVED** [review]

**Pull request merged**